Refine QS for manually connecting Node.js (#227)
redhat-developer/app-services-guides: nodejs/README.adoc

* https://github.com/git-guides/[Git^] is installed.
* You have an IDE such as https://www.jetbrains.com/idea/download/[IntelliJ IDEA^], https://www.eclipse.org/downloads/[Eclipse^], or https://code.visualstudio.com/Download[VSCode^].
* https://nodejs.org/en/download/[Node.js 14.x^] is installed.

[NOTE]
====
The example Node.js application in this quick start uses a https://kafka.js.org/[KafkaJS^] client by default. If you want to use a https://github.com/blizzard/node-rdkafka[node-rdkafka^] client, you must install Docker to run a specified container image and configure a development environment. To learn more, see the https://github.com/nodeshift-starters/reactive-example/blob/main/README.md[documentation] for the Node.js application.
====

// Condition out QS-only content so that it doesn't appear in docs.
// All QS anchor IDs must be in this alternate anchor ID format `[#anchor-id]` because the ascii splitter relies on the other format `[id="anchor-id"]` to generate module files.
ifdef::qs[]
[#description]
Learn how to use Node.js applications to produce and consume messages using a Kafka instance in {product-long}.

[#introduction]

> Welcome to the quick start for {product-long} with Node.js. In this quick start, you'll learn how to use https://nodejs.org/en/about/[Node.js^] to produce messages to and consume messages from your Kafka instances in {product}.

Reviewer: use Node.js sounds strange here. I would rephrase it to something like: "learn to use Node.js runtime to..

endif::[]

[id="proc-importing-nodejs-sample-code_{context}"]
== Importing the Node.js sample code

For this quick start, you'll use sample code from the Nodeshift Application Starters https://github.com/nodeshift-starters/reactive-example[reactive-example^] repository in GitHub. After you understand the concepts and tasks in this quick start, you can use your own Node.js applications with {product} in the same way.

.Procedure
. On your computer, open a command-line window.

> . On the command line, clone the Nodeshift Application Starters https://github.com/nodeshift-starters/reactive-example[reactive-example^] repository from GitHub.

Reviewer: ```suggestion
. On the command line, clone the Nodeshift Application Starter https://github.com/nodeshift-starters/reactive-example[reactive-example^] repository from GitHub.
```

Reviewer: No strong opinion. I just found it odd when reading.
That name reflects organization. What we cloning is repository which is instance of single starter.

Author: Yes, sorry, _organization_ was the word I was looking for - not project. Regardless, I think we are right to name it as it appears.

+
.Cloning the reactive-example repository
[source,subs="+attributes"]
----
$ git clone https://github.com/nodeshift-starters/reactive-example.git
----
. In your IDE, open the `reactive-example` directory corresponding to the repository that you cloned.

ifdef::qs[]
.Verification
* Is the Node.js example application accessible in your IDE?
endif::[]

[id="proc-configuring-nodejs_{context}"]
== Configuring the Node.js example application to connect to a Kafka instance

To enable your Node.js application to access a Kafka instance, you need to configure a connection by specifying the following details:

* The bootstrap server endpoint for your Kafka instance
* The generated credentials for your {product} service account
* The Simple Authentication and Security Layer (SASL) mechanism that the client will use to authenticate with the Kafka instance

> In this task, you'll create a new shell script called `set-environment.sh`. In this shell script, you'll set the required bootstrap server and client credentials as environment variables.

Reviewer 2: ```
In this task, you'll create a new file called `.env`. In this file you'll set the required bootstrap server and client credentials so that they can be used to connect to your Kafka instance.
```

.Prerequisites
ifndef::qs[]
* You have the bootstrap server endpoint and client credentials for the Kafka instance. You can access this information as follows:
** For the bootstrap server endpoint, select the options menu (three vertical dots) for a Kafka instance in {product}. Click *View connection information*.
** For the client ID and client secret, you're prompted to copy and store these credentials when you create a service account in {product}.
endif::[]

.Procedure

> . In your IDE, open a new file for the shell script. Save the file with the name `set-environment.sh`.

Reviewer 2: ```suggestion
. In your IDE, open a new file for the shell script. Save the file with the name `.env`.
```

> . In the `set-environment.sh` file, add the lines shown in the example to set the bootstrap server and client credentials as environment variables to be used by the Node.js application.

Reviewer 2: ```suggestion
. In the `.env` file, add the lines shown in the example to set the bootstrap server and client credentials as environment variables to be used by the Node.js application.
```

+
.Setting environment variables for bootstrap server and client credentials
[source,subs="+quotes"]
----

> export KAFKA_BOOTSTRAP_SERVER=__<bootstrap_server>__
> export KAFKA_CLIENT_ID=__<client_id>__
> export KAFKA_CLIENT_SECRET=__<client_secret>__
> export KAFKA_SASL_MECHANISM=plain

Reviewer 2: ```suggestion
KAFKA_BOOTSTRAP_SERVER=__<bootstrap_server>__
KAFKA_CLIENT_ID=__<client_id>__
KAFKA_CLIENT_SECRET=__<client_secret>__
KAFKA_SASL_MECHANISM=plain
```

----
+
In the preceding example, replace the values in angle brackets (`< >`) with your own bootstrap server and client credential information.
ifdef::qs[]
These details are described below.
+
--
<bootstrap_server>:: The bootstrap server endpoint for your service account. To access this information for a Kafka instance in {product}, select the options menu (three vertical dots). Click *View connection information*.
<client_id>:: A client credential generated when you create a service account in {product}. You're prompted to copy and store this credential when you create the service account.
<client_secret>:: A client credential generated when you create a service account in {product}. You're prompted to copy and store this credential when you create the service account.
--
endif::[]
+
In this case, observe that the Node.js application uses the `SASL/PLAIN` authentication method (that is, the value of `KAFKA_SASL_MECHANISM` is set to `plain`). This means that the application uses only the client ID and client secret to authenticate with the Kafka instance. The application *doesn't* also require an authentication token.

> . Save the the `set-environment.sh` file.

Reviewer 2: ```suggestion
. Save the the `.env` file.
```

ifdef::qs[]
.Verification
* Did you set environment variables for the Kafka instance?
endif::[]

[id="proc-creating-countries-topic_{context}"]
== Creating a Kafka topic in {product}

In this quick start, the Node.js example application requires a Kafka topic called `countries`. You need to create this topic in your Kafka instance so that the application can interact with it.

.Prerequisites
* You've created a Kafka instance in {product} and the instance is in the *Ready* state.

.Procedure
. In the {product} web console, go to *Streams for Apache Kafka* > *Kafka Instances* and click the name of the Kafka instance that you want to add a topic to.
. Click *Create topic* and follow the guided steps to define the topic details. Click *Next* to complete each step and click *Finish* to complete the setup.
+
[.screencapture]

> .Guided steps to define topic details

Reviewer: This repeats a lot with other guides. Worth checking if content is the same

Reviewer 2: The version in the doc I created was copied from the Quarkus guide, except for the topic create.

Author: Yes, we are using the same content as for other quick starts (for example, Quarkus), but with some minor tweaks.

.Verification
ifdef::qs[]
* Is the `countries` topic listed in the topics table?
endif::[]
ifndef::qs[]
* Verify that the `countries` topic is listed in the topics table.
endif::[]

[id="proc-running-nodejs-example-application_{context}"]
== Running the Node.js example application

After you configure your Node.js application to connect to a Kafka instance and you create the required Kafka topic, you can run the Node.js application to start producing and consuming messages.

In this task, you'll use the following components of the Node.js application to produce and consume messages:

* A `producer-backend` component that generates random country names and sends these names to the Kafka topic.
* A `consumer-backend` component that consumes the country names from the Kafka topic.

.Prerequisites
* You've configured the Node.js example application to connect to the Kafka instance.
* You've created the `countries` example Kafka topic.

.Procedure
. On the command line, navigate to the `reactive-example` directory corresponding to the repository that you cloned.
+
.Navigating to the reactive-example directory
[source]
----
$ cd ~/reactive-example
----

> . Check out the `kafkajs` branch of the repository.

Reviewer: I think this is the default branch right now when we checkout source code is it might not be needed.

WDYT @mhdawson ?

Ref:
https://github.com/nodeshift-starters/reactive-example

Reviewer 2: I had just changed that recently I think @jbyrne-redhat was planning to remove it.

Author: Yes, given that `kafkajs` is now the default branch, I've removed the step that explicitly instructs users to check it out.

+
.Checking out the kafkajs branch
[source]
----
$ git checkout kafkajs
----

> . Run the shell script that sets environment variables for the bootstrap server endpoint and client credentials. These are required to connect to your Kafka instance.
> +
> .Running the shell script
> [source]
> ----
> $ source set-environment.sh
> ----
> 

Reviewer 2: ```suggestion
```

+
.Installing the consumer component
[source]
----
$ cd consumer-backend
$ npm install
----

. Run the consumer component.
+
.Running the consumer component
[source]
----
$ node consumer.js
----
+
You should see the Node.js application start to run and connect to the Kafka instance. However, because you haven't run the producer component yet, the consumer has no country names to display.
+
If the application fails to run, review the error log in the command-line window and address any problems. Also, review the steps in this quick start to ensure that the application and Kafka topic are configured correctly.

. On your computer, open a second command-line window or tab. On the command line, navigate to the `reactive-example` directory corresponding to the repository that you cloned.
+
.Navigating to the reactive-example directory
[source]
----
$ cd ~/reactive-example
----

. Check out the `kafkajs` branch of the repository.
+
.Checking out the kafkajs branch
[source]
----
$ git checkout kafkajs
----

. Run the shell script that sets environment variables for the bootstrap server endpoint and client credentials.
+
.Running the shell script
[source]
----

> $ source set-environment.sh

Reviewer: This is antipattern for node.js.

Would it be possible to use something like https://www.npmjs.com/package/dotenv in the repo and remove those steps from guide?

Reviewer 2: I did that to avoid having people having to export the variables multiple times while still going over what you need to do so that people would understand and can apply to whatever they will do in their environment.  I looks like using a .env file might be similar.  They key thing for me is that the person going through it knows what they need to set/use in their code.

Reviewer 2: The nice thing about using .env would be that it should work the same across platforms so for that reason I think it's worth doing.

I'll see if I can find tomorrow to adjust for using .env.

Reviewer: Thank you. .Env is used by Quarkus guide as well.

Author: @mhdawson  Let me know if you implement this change. Right now, we instruct users to create a shell script for Linux/macOS, or a batch file for Windows.

Reviewer 2: @jbyrne-redhat will do, pretty busy but might find time on Friday , will let you know.

Reviewer 2: From https://access.redhat.com/documentation/en-us/red_hat_openshift_streams_for_apache_kafka/1/guide/04827d87-ed92-4ffd-a126-11fa13348eba looks to me like the quarkus guide uses environment variables, but still think its a good idea to look at the .env approach for the Node.js guide.

Reviewer 2: PR to update the kafka version to support dotenv: https://github.com/nodeshift-starters/reactive-example/pull/34. Working on node-rdkafka version.

Reviewer 2: PR for node-rdkafka - https://github.com/nodeshift-starters/reactive-example/pull/35

Both PRs are landed, and with the suggestions I made to swtich to creating '.env' file I think I'm done.

@jbyrne-redhat let me know if you need anything else.

> . Check out the `kafkajs` branch of the repository.
> +
> .Checking out the kafkajs branch
> [source]
> ----
> $ git checkout kafkajs
> ----
> 
> . Run the shell script that sets environment variables for the bootstrap server endpoint and client credentials.
> +
> .Running the shell script
> [source]
> ----
> $ source set-environment.sh

Reviewer 2: ```suggestion
```

> ----

Reviewer 2: ```suggestion
```

. Navigate to the directory for the producer component. Use Node Package Manager (npm) to install the package for this component.
+
.Installing the producer component
[source]
----

> $ cd producer-backend

Reviewer: While those steps are easy it felt that we could have them grouped together with some script on the repository.
For example npm workspace can run multiple projects at the same time with the proper logs.

Reviewer 2: I sometimes find that when its just run this script then that script you don't see what is going on. Since these are quite simple I"d lean towards leaving the individual steps. They don't take long and you see the exact steps.

Reviewer: Agreed.

+
.Example output from the producer component
[source]
----
$ node producer.js
Ghana
Réunion
Guatemala
Luxembourg
Mayotte
Syria
United Kingdom
Bolivia
Haiti
----
+
As shown in the example, the producer component starts to run and generate messages that represent country names. The producer writes all messages to the single partition that you specified when you created your Kafka topic.

. Switch back to the first command-line window that you opened.
+
You should now see that the consumer component displays the same country names generated by the producer, and in the same order, as shown in the following example:
+
.Example output from the consumer component
[source]
----
$ node consumer.js
Ghana
Réunion
Guatemala
Luxembourg
Mayotte
Syria
United Kingdom
Bolivia
Haiti
----
+
The output from both components shows that the Node.js application successfully connected to the Kafka instance. The application is using the Kafka topic you created to produce and consume messages.

. To verify __how__ the Node.js application connected to the Kafka instance, perform the following actions:
.. In your IDE, open the `producer.js` file.
+

> Observe that the producer component is using the environment variables that you set in the `set-environment.sh` shell script. The values of these environment variables are the bootstrap server endpoint and client credentials that the producer component used to connect to the Kafka instance.

Reviewer 2: ```suggestion
Observe that the producer component is using the environment variables that you set in the `.env` file. The values of these environment variables are the bootstrap server endpoint and client credentials that the producer component used to connect to the Kafka instance.

```

Author: Thanks a lot for making these updates, @mhdawson. I made some other updates on my branch in the meantime, so I'll incorporate your updates also. I should be able to push all those changes sometime Monday and will ask you for a final review then.
